### PR TITLE
[core] Replace task resubmission in raylet with ownership protocol

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Currently, the raylet handles plasma objects whose values cannot be retrieved by setting a timeout (`initial_reconstruction_timeout_milliseconds`). After the timeout passes, the raylet attempts to resubmit the task and fails the object if the task spec is not available. This was part of the legacy raylet-based architecture and is unstable (if the timeout is set too low, an error can be spuriously thrown) and incompatible with ownership-based object reconstruction (since the owner will now resubmit tasks instead of the raylet).

This PR updates the failure handling to contact the owner instead. The raylet will no longer resubmit tasks; instead, after the timeout is triggered, the raylet contacts the owner to check that the owner is still alive and still has the object in reference. If true, the raylet does nothing and lets the owner manage the object. If false, the raylet stores an error in plasma that gets thrown to the receiving worker, as before.

This PR also modifies the protocol for resolving borrowed ObjectIDs by storing an error if the owner has died. This is not necessary for correctness but reduces the time needed to detect the failure.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by edoakes** [comment]
Looks great @stephanie-wang! Is it worth doing any benchmarking on this? From my understanding the only additional overhead should be on the owner once we reach the fetch timeout, which should be a rare case, so maybe it's unnecessary.

**Comment by stephanie-wang** [comment]
> Looks great @stephanie-wang! Is it worth doing any benchmarking on this? From my understanding the only additional overhead should be on the owner once we reach the fetch timeout, which should be a rare case, so maybe it's unnecessary.

Hmm yeah, I didn't do any benchmarking for the reason that you said. So far I've just run the relevant Python tests locally and didn't notice any significantly different runtimes. But let me know if you think of something I should try running!

In the future, we can also speed up the failure detection even more by having the raylet fail objects as soon as an owner fails (see the TODO in this PR).

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1682,31 +1690,32 @@ void CoreWorker::HandleGetObjectStatus(const rpc::GetObjectStatusRequest &reques
...
-      RemoveLocalReference(object_id);
-    } else {
-      // We lost the race, the task is done.
-      RemoveLocalReference(object_id);
-      send_reply_callback(Status::OK(), nullptr, nullptr);
-    }
-  } else {
-    // The task is done. Send the reply immediately.
+  // Acquire a reference to the object. This prevents the object from being
+  // evicted out from under us while we check the object status and start the
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc:1721]: Oh yeah, it's a bit strange, but the callbacks will get called on `Put` even if the ref count is 0. So we just need to make sure that the callback gets added either 1) while the object is in the in-memory store, or 2) before the object has been `Put` in the in-memory store.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -3000,76 +3001,120 @@ void NodeManager::FinishAssignedActorCreationTask(const ActorID &parent_actor_id
...
+          // object manager will continue trying to fetch the object, and this
+          // handler will get triggered again if the object is still
+          // unavailable after another timeout.
+        }));
+  } else {
+    // We do not have the owner's address. This is either an actor creation
+    // task or a randomly generated ObjectID. Try to look up the spec for the
+    // actor creation task.
+    // TODO(swang): The task lookup is only needed when the GCS actor service is
+    // disabled. Once the GCS actor service is enabled by default, we can
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:3049]: I'm wondering if we should just not implement this given that GCS actor service is coming soon, and this is flagged off by default?

Or do we rely on this for the normal actor reconstruction path?

**Review by ericl — APPROVED** [review]

File: src/ray/raylet/node_manager.cc (hunk 2 above)
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:3049]: It's to make sure that actor reconstruction works when the GCS actor service is flagged off. We can remove it once the GCS actor service is the default.

**Pull request merged**